File: 0287.FindtheDuplicateNumber.py
class Solution(object):
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) <= 1:
            return
        slow = nums[0]
        fast = nums[nums[0]]
        while slow != fast:
            slow = nums[slow]
            fast = nums[nums[fast]]
        slow = 0
        while slow != fast:
            slow = nums[slow]
            fast = nums[fast]
        return slow
        
        '''
        dic = {}
        for num in nums:
            dic[num] = dic.get(num, 0) + 1
        for num in dic.keys():
            if dic[num] != 1:
                return num
        '''    

File: 0000-0999/0287.FindtheDuplicateNumber.go
func findDuplicate(nums []int) int {
    // No modify, O(n) + O(1)
    slow, fast := nums[0], nums[nums[0]]
    for slow != fast {
        slow, fast = nums[slow], nums[nums[fast]]
    }
    slow2 := 0
    for slow != slow2 {
        slow, slow2 = nums[slow], nums[slow2]
    }
    return slow
    /* Swap, O(n) + O(1)
    for i := range nums {
        for nums[i] != nums[nums[i]-1] {
            nums[i], nums[nums[i]-1] = nums[nums[i]-1], nums[i]
        }
    }
    for i, num := range nums {
        if num != i+1 {
            return num
        }
    }
    return -1
    */
}